compiler/codegen/CodeGenPrep.cpp

```diff
@@ -177,6 +179,128 @@ OMR::CodeGenerator::lowerTreesPreChildrenVisit(TR::Node * parent, TR::TreeTop *
+ *
+ *    treetop
+ *      fload x
+ *    treetop
+ *      fload y
+ *    treetop
+ *      fload z
+ *    treetop
+ *      new jitNewValueObject (skipZeroInit)
+ *        loadaddr Vec3
+ *    istorei Vec3.x
+ *      ==>new
+ *      ==>fload x
+ *    istorei Vec3.y
+ *      ==>new
+ *      ==>fload y
+ *    istorei Vec3.z
+ *      ==>new
+ *      ==>fload z
+ *    fence
+ *
+ * Note that the (helper) symbol reference for the `new` node is be
+ * the same as the symbol reference for the `newvalue` node.
+ *
+ * The fence at the end is needed for platforms with weak memory ordering
+ * such as POWER. It may appear anywhere after allocation + initialization
+ * but before any operation that could publish the resulting reference to
+ * other threads.
+ *
+ * @param comp pointer to the compilation object
+ * @param node the node being lowered
+ * @param tt the TreeTop anchoring the node
+ */
+static void
+lowerNewValue(TR::Compilation *comp, TR::Node *node, TR::TreeTop *tt)
+   {
+   auto* prevTreeTop = tt->getPrevTreeTop();
+
+   // remove tt from the TreeTop List
+   prevTreeTop->setNextTreeTop(NULL);
+
+   // Iterate over every child that sets a field. The first
+   // child is skipped as it's only used to specify the type
+   // of the value being constructed. The second child is
+   // the one that actually initializes the first field.
+   for (int i = 1; i < node->getNumChildren(); ++i)
+      {
+      auto* ttNode = TR::Node::create(TR::treetop, 1);
+      ttNode->setFirst(node->getChild(i));
+      prevTreeTop = TR::TreeTop::create(comp, ttNode, NULL, prevTreeTop);
+      prevTreeTop->getPrevTreeTop()->setNextTreeTop(prevTreeTop);
+      }
+
+   // Create new TreeTop to anchor the new. The TreeTop for
+   // the newValue cannot be reused because it is still used
+   // just after lowering the node.
+   auto* newNode = TR::Node::createWithSymRef(TR::New, 1, node->getSymbolReference());
+   newNode->setFirst(node->getChild(0));
+   auto* ttNode = TR::Node::create(TR::treetop, 1);
+   ttNode->setAndIncChild(0, newNode);
+   prevTreeTop = TR::TreeTop::create(comp, ttNode, NULL, prevTreeTop);
+   prevTreeTop->getPrevTreeTop()->setNextTreeTop(prevTreeTop);
+
+   // Get the class and layout of the value type
+   auto* valueClass = static_cast<TR_OpaqueClassBlock *>(static_cast<TR::StaticSymbol *>(newNode->getSymbol())->getStaticAddress());
+   const TR::TypeLayout* typeLayout = comp->typeLayout(valueClass);
+
+   // Iterate over every child that sets a field and generated
+   // a store for the value.
+   for (int i = 1; i < node->getNumChildren(); ++i)
+      {
+      TR::IL il;
+      auto storeOpCode = il.opCodeForIndirectStore(node->getChild(i)->getDataType());
+      const TR::TypeLayoutEntry fieldEntry = typeLayout->entry(i - 1);
+      const bool isVolatile = false; // fields of newvalue allocated objects are,
+      const bool isPrivate = false;  // by definition, non-volatile, non-private,
+      const bool isFinal = true;     // and final
```

Comment:
I'm not sure if this comment is correct.

I believe it reflects the original intention/motivation for the `newvalue` opcode, but I'm not sure if it's generalizable. Wouldn't it be possible for a consumer of the OMR compiler to use `newvalue` but have different expectations for `isVolatile`, `isPrivate`, and `isFinal`?

Reply:
I think this needs to be made an object model query so that the various languages can specify the behaviour.

Reply:
I've added these fields to the TypeLayout class so that downstream projects can now set appropriate values for each.